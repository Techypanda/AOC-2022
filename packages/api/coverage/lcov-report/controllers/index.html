
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for controllers</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../index.html">All files</a> controllers</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">71.81% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>186/259</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">61.73% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>71/115</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">81.13% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>43/53</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">72.56% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>164/226</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="Calories.ts"><a href="Calories.ts.html">Calories.ts</a></td>
	<td data-value="78.94" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 78%"></div><div class="cover-empty" style="width: 22%"></div></div>
	</td>
	<td data-value="78.94" class="pct medium">78.94%</td>
	<td data-value="19" class="abs medium">15/19</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="6" class="abs medium">4/6</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="5" class="abs high">4/5</td>
	<td data-value="76.47" class="pct medium">76.47%</td>
	<td data-value="17" class="abs medium">13/17</td>
	</tr>

<tr>
	<td class="file high" data-value="Cleaning.ts"><a href="Cleaning.ts.html">Cleaning.ts</a></td>
	<td data-value="83.87" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 83%"></div><div class="cover-empty" style="width: 17%"></div></div>
	</td>
	<td data-value="83.87" class="pct high">83.87%</td>
	<td data-value="31" class="abs high">26/31</td>
	<td data-value="86.11" class="pct high">86.11%</td>
	<td data-value="36" class="abs high">31/36</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	<td data-value="85.71" class="pct high">85.71%</td>
	<td data-value="28" class="abs high">24/28</td>
	</tr>

<tr>
	<td class="file high" data-value="Jajanken.ts"><a href="Jajanken.ts.html">Jajanken.ts</a></td>
	<td data-value="89.74" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 89%"></div><div class="cover-empty" style="width: 11%"></div></div>
	</td>
	<td data-value="89.74" class="pct high">89.74%</td>
	<td data-value="39" class="abs high">35/39</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="6" class="abs high">5/6</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	<td data-value="88.88" class="pct high">88.88%</td>
	<td data-value="36" class="abs high">32/36</td>
	</tr>

<tr>
	<td class="file medium" data-value="Rucksack.ts"><a href="Rucksack.ts.html">Rucksack.ts</a></td>
	<td data-value="58.92" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 58%"></div><div class="cover-empty" style="width: 42%"></div></div>
	</td>
	<td data-value="58.92" class="pct medium">58.92%</td>
	<td data-value="112" class="abs medium">66/112</td>
	<td data-value="40" class="pct low">40%</td>
	<td data-value="55" class="abs low">22/55</td>
	<td data-value="78.26" class="pct medium">78.26%</td>
	<td data-value="23" class="abs medium">18/23</td>
	<td data-value="59.13" class="pct medium">59.13%</td>
	<td data-value="93" class="abs medium">55/93</td>
	</tr>

<tr>
	<td class="file medium" data-value="Stacker.ts"><a href="Stacker.ts.html">Stacker.ts</a></td>
	<td data-value="75.86" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 75%"></div><div class="cover-empty" style="width: 25%"></div></div>
	</td>
	<td data-value="75.86" class="pct medium">75.86%</td>
	<td data-value="58" class="abs medium">44/58</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="12" class="abs medium">9/12</td>
	<td data-value="77.77" class="pct medium">77.77%</td>
	<td data-value="9" class="abs medium">7/9</td>
	<td data-value="76.92" class="pct medium">76.92%</td>
	<td data-value="52" class="abs medium">40/52</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2022-12-05T09:11:23.883Z
            </div>
        <script src="../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../sorter.js"></script>
        <script src="../block-navigation.js"></script>
    </body>
</html>
    