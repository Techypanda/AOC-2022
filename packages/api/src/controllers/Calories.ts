import { FastifyReply } from 'fastify';
import { logger } from '../utils/logger';
interface CaloriesRequest {
    data: string[]
}
export class CaloriesController {
    async calculateTopNCalories(input: string[], n = 1) {
        let current = 0;
        const sums: number[] = [];
        input.forEach((x, idx) => {
            if (x === '' || idx === input.length-1) {
                sums.push(current);
                current = 0;
            } else {
                current += Number.parseFloat(x);
            }
        });
        const sorted = sums.sort((a, b) => a - b);
        const max = sorted.slice(n * -1);
        const sum = max.reduce((prev, curr) => prev + curr, 0);
        return { max, sum };
    }
    async get(request: CaloriesRequest, reply: FastifyReply, n = 1) {
        logger.info(`Calculating top ${n} calories`);
        const { max, sum } = await this.calculateTopNCalories(request.data, n);
        logger.info('finished calculating top n calories', { max, sum });
        reply
            .status(200)
            .header('Content-Type', 'application/json; charset=utf-8')
            .send({ max, sum });
    }
}