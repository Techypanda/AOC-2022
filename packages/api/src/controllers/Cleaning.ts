import { FastifyReply } from "fastify";
import { logger } from "../utils/logger";

// map abuse again i guess
type MinMaxPair = [number, number]; // Min, Max
export enum OverlapStatus {
    OverlapsEntirely,
    Overlaps,
    DoesNot
}
export interface CleaningResponse {
    overlapStatuses: OverlapStatus[];
    sum: number;
}
interface CleaningRequest {
    data: string[];
    v2: boolean;
}
export class Cleaning {
    parseLine(line: string): [MinMaxPair, MinMaxPair] {
        const pairs = line.split(',');
        const resp = pairs.map((pair) => pair.split('-').map((str) => Number.parseInt(str)));
        return resp as unknown as [MinMaxPair, MinMaxPair];
    }
    overlaps(p1: MinMaxPair, p2: MinMaxPair): OverlapStatus {
        if (p1.length < 2 || p2.length < 2) return OverlapStatus.DoesNot;
        // Ternary Hell
        const OE_CASE_1 = p1[0] >= p2[0] && p1[1] <= p2[1];
        const OE_CASE_2 = p2[0] >= p1[0] && p2[1] <= p1[1];
        // 0, 3, 3, 4 : 0 <= 3 && 3 >= 3 i.e slightly behind
        const O_CASE_1 = p1[0] < p2[0] && p1[1] >= p2[0];
        const O_CASE_2 = p2[0] < p1[0] && p2[1] >= p1[0];
        // 3, 6, 0, 3 : 3 <= 3 && 6 >= 3 i.e slightly ahead
        const O_CASE_3 = p1[0] <= p2[1] && p1[1] >= p2[1];
        const O_CASE_4 = p2[0] <= p1[1] && p2[1] >= p1[1];
        return OE_CASE_1 ? OverlapStatus.OverlapsEntirely // Case 1: P1 INSIDE P2
            : OE_CASE_2 ? OverlapStatus.OverlapsEntirely // Case 2: P2 INSIDE P1
            : O_CASE_1 ? OverlapStatus.Overlaps // Case 3: P1 Overlaps from MIN but goes past p2
            : O_CASE_2 ? OverlapStatus.Overlaps // Case 4: P2 overlaps from MIN but goes past p2
            : O_CASE_3 ? OverlapStatus.Overlaps // Case 5: P1 Overlaps but MIn slightly less, however max is in bounds
            : O_CASE_4 ? OverlapStatus.Overlaps // Case 6: P2 Overlaps but MIn slightly less, however max is in bounds
            : OverlapStatus.DoesNot;
    }
    parseInput(lines: string[], conditionToWatch: OverlapStatus = OverlapStatus.OverlapsEntirely): CleaningResponse {
        let sum = 0;
        const overlapStatuses = lines.map((x) => {
            const parsedLine = this.parseLine(x);
            const overLapStatus = this.overlaps(parsedLine[0], parsedLine[1]);
            sum += overLapStatus === conditionToWatch || overLapStatus === OverlapStatus.OverlapsEntirely ? 1 : 0; // faster then doing another iteration over output array to calculate this
            return overLapStatus;
        });
        return { overlapStatuses, sum };
    }
    async post(request: CleaningRequest, reply: FastifyReply) {
        logger.info('Calculating cleaning overlaps');
        const resp = this.parseInput(request.data, request.v2 ? OverlapStatus.Overlaps : OverlapStatus.OverlapsEntirely);
        logger.info('calculated cleaning overlaps', resp);
        reply
            .status(200)
            .header('Content-Type', 'application/json; charset=utf-8')
            .send(resp);
    }
}