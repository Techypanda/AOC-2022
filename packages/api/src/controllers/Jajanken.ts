import { FastifyReply } from "fastify";
import { logger } from "../utils/logger";

// https://japaneseanime.fandom.com/wiki/Gon_Freecss#Jajanken_Scissors
enum JajankenMoves {
    Rock = 1,
    Paper = 2,
    Scissors = 3
}
type JajankenScore = Record<JajankenMoves, number>;
type ScoreJajanken = Record<number, JajankenMoves>; // flipped
type JajankenInputs = 'A' | 'B' | 'C' | 'X' | 'Y' | 'Z';
const RockScore: JajankenScore = {
    [JajankenMoves.Rock]: 3,
    [JajankenMoves.Paper]: 0,
    [JajankenMoves.Scissors]: 6
}
const ScoreRock = Object.fromEntries(Object.entries(RockScore).map(([key, value]) => [value, key])) as unknown as ScoreJajanken;
const ScissorsScore: JajankenScore = {
    [JajankenMoves.Rock]: 0,
    [JajankenMoves.Paper]: 6,
    [JajankenMoves.Scissors]: 3
}
const ScoreScissors = Object.fromEntries(Object.entries(ScissorsScore).map(([key, value]) => [value, key])) as unknown as ScoreJajanken;
const PaperScore: JajankenScore = {
    [JajankenMoves.Rock]: 6,
    [JajankenMoves.Paper]: 3,
    [JajankenMoves.Scissors]: 0
}
const ScorePaper = Object.fromEntries(Object.entries(PaperScore).map(([key, value]) => [value, key])) as unknown as ScoreJajanken;
const Score: Record<JajankenMoves, JajankenScore> = {
    [JajankenMoves.Rock]: RockScore,
    [JajankenMoves.Paper]: PaperScore,
    [JajankenMoves.Scissors]: ScissorsScore
}
const ScoreOpposite: Record<JajankenMoves, ScoreJajanken> = {
    [JajankenMoves.Rock]: ScoreRock,
    [JajankenMoves.Paper]: ScorePaper,
    [JajankenMoves.Scissors]: ScoreScissors
}
const StrategyGuide: Record<JajankenInputs, JajankenMoves> = {
    A: JajankenMoves.Rock,
    B: JajankenMoves.Paper,
    C: JajankenMoves.Scissors,
    X: JajankenMoves.Rock,
    Y: JajankenMoves.Paper,
    Z: JajankenMoves.Scissors
}
const StategyGuideV2: Record<'X' | 'Y' | 'Z', number> = {
    X: 0,
    Y: 3,
    Z: 6
}
interface JanjankenGameResponse {
    scores: number[];
    sum: number;
}
interface JanjankenRequest {
    data: string[];
    v2: boolean;
}
export class Jajanken {
    playRound(opposition: JajankenInputs, player: JajankenInputs, v2 = false) {
        if (v2) {
            const chosenByOpponent = StrategyGuide[opposition];
            const pointsNeeded = 6 - StategyGuideV2[player as 'X' | 'Y' | 'Z'];
            const scoreValue = ScoreOpposite[chosenByOpponent][pointsNeeded];
            const score = Number.parseInt(StategyGuideV2[player as 'X' | 'Y' | 'Z']  as unknown as string) + Number.parseInt(scoreValue as unknown as string); // typescript madness happened
            return score;
        }
        return Score[StrategyGuide[player]][StrategyGuide[opposition]] + StrategyGuide[player];
    }
    async runGame(input: string[], v2 = false): Promise<JanjankenGameResponse> {
        const scores: number[] = [];
        let sum = 0;
        input.forEach((line) => {
            const plays = line.split(' ') as JajankenInputs[];
            if (plays.length === 2) {
                const round = this.playRound(plays[0], plays[1], v2);
                scores.push(round);
                sum += round;
            }
        });
        return { scores, sum };
    }
    async post(request: JanjankenRequest, reply: FastifyReply) {
        logger.info('Calculating jajanken game');
        const { scores, sum } = await this.runGame(request.data, request.v2);
        logger.info('finished jajanken game', { scores, sum });
        reply
            .status(200)
            .header('Content-Type', 'application/json; charset=utf-8')
            .send({ scores, sum });
    }
}