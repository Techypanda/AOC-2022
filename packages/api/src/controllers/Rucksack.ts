import { FastifyReply } from "fastify";
import { logger } from "../utils/logger";

interface RucksackRequest {
    data: string[];
    v2: boolean;
}

export class Rucksack {
    getCompartments(input: string) {
        const compartmentOne = input.substring(0, input.length/2);
        const compartmentTwo = input.substring(input.length/2, input.length);
        return { compartmentOne, compartmentTwo };
    }
    private doublePointerTrace(compartmentOne: string[], compartmentTwo: string[]): string[] { // O(N) algorithm
        const sortedCompartmentOne = compartmentOne.sort((x, y) => x.charCodeAt(0) - y.charCodeAt(0));
        const sortedCompartmentTwo = compartmentTwo.sort((x, y) => x.charCodeAt(0) - y.charCodeAt(0));
        let i = 0;
        let k = 0;
        const existsInBoth: string[] = [];
        while (i < compartmentOne.length && k < compartmentTwo.length) {
            // logger.info({ iAt: sortedCompartmentOne[i], kAt: sortedCompartmentTwo[k], i, k, iAscii: sortedCompartmentOne[i].charCodeAt(0), kAscii: sortedCompartmentTwo[k].charCodeAt(0) })
            if (sortedCompartmentOne[i] === sortedCompartmentTwo[k]) {
                // SAME
                existsInBoth.push(sortedCompartmentOne[i]);
                let tmpI = i;
                let tmpK = k;
                while (true && i < compartmentOne.length && k < compartmentTwo.length) {
                    if (sortedCompartmentOne[tmpI] === sortedCompartmentOne[i]) {
                        tmpI++;
                    } else if (sortedCompartmentTwo[tmpK] === sortedCompartmentTwo[k]) {
                        tmpK++;
                    } else {
                        break;
                    }
                }
                i = tmpI; k = tmpK;
            } else if (sortedCompartmentOne[i].charCodeAt(0) < sortedCompartmentTwo[k].charCodeAt(0)) {
                i++;
            } else if (sortedCompartmentTwo[k].charCodeAt(0) < sortedCompartmentOne[i].charCodeAt(0)) {
                k++;
            } else {
                i++; k++;
            }
        }
        return existsInBoth;
    }
    private triplePointerTrace(c1: string[], c2: string[], c3: string[]): string[] { // could combine ^ and this with recursion but im lazy and its 9pm
        const sortedC1 = c1.sort((x, y) => x.charCodeAt(0) - y.charCodeAt(0));
        const sortedC2 = c2.sort((x, y) => x.charCodeAt(0) - y.charCodeAt(0));
        const sortedC3 = c3.sort((x, y) => x.charCodeAt(0) - y.charCodeAt(0));
        let i = 0;
        let k = 0;
        let j = 0;
        const existsInBoth: string[] = [];
        while (i < c1.length && k < c2.length && j < c3.length) {
            if (sortedC1[i] === sortedC2[k] && sortedC2[k] === sortedC3[j]) {
                existsInBoth.push(sortedC1[i]);
                let tmpI = i;
                let tmpK = k;
                let tmpJ = j;
                while (true && i < c1.length && k < c2.length && j < c3.length) {
                    if (sortedC1[tmpI] === sortedC1[i]) {
                        tmpI++;
                    } else if (sortedC2[tmpK] === sortedC2[k]) {
                        tmpK++;
                    } else if (sortedC3[tmpJ] === sortedC3[j]) {
                        tmpJ++;  
                    } else {
                        break;
                    }
                }
                i = tmpI; k = tmpK; j = tmpJ;
            } else if (sortedC1[i].charCodeAt(0) < sortedC2[k].charCodeAt(0) && sortedC1[i].charCodeAt(0) < sortedC3[j].charCodeAt(0)) { // i is smallest
                i++;
            } else if (sortedC2[k].charCodeAt(0) < sortedC1[i].charCodeAt(0) && sortedC2[k].charCodeAt(0) < sortedC3[j].charCodeAt(0)) { // k is smallest
                k++;
            } else if (sortedC3[j].charCodeAt(0) < sortedC1[i].charCodeAt(0) && sortedC3[j].charCodeAt(0) < sortedC2[k].charCodeAt(0)) { // j is smallest
                j++;
            } else {
                i++; j++; k++;
            }
        }
        return existsInBoth;
    }
    private asciiToPriority(n: number) {
        return n < 91 ? n - 38 : n - 96;
    }
    processLine(input: string) {
        const compartments = this.getCompartments(input);
        const existsInBoth = this.doublePointerTrace(compartments.compartmentOne.split(''), compartments.compartmentTwo.split(''));
        const sum = existsInBoth.reduce((prev, curr) => prev + this.asciiToPriority(curr.charCodeAt(0)), 0);
        return { existsInBoth, sum };
    }
    parseText(input: string[]) {
        let sum = 0;
        input.forEach((x) => sum += this.processLine(x).sum);
        return sum;
    }
    parseGroup(input: string[], groups = 3) {
        let sum = 0;
        for (let i = 0; i + 2 < input.length; i+=3) {
            const m: Record<string, number> = {};
            const filteredOne = input[i].split('').filter((v, i, s) => s.indexOf(v) === i);
            const filteredTwo = input[i+1].split('').filter((v, i, s) => s.indexOf(v) === i);
            const filteredThree = input[i+2].split('').filter((v, i, s) => s.indexOf(v) === i);
            filteredOne.map((x) => {
                if (m[x]) {
                    m[x] += 1;
                } else {
                    m[x] = 1;
                }
            });
            filteredTwo.map((x) => {
                if (m[x]) {
                    m[x] += 1;
                } else {
                    m[x] = 1;
                }
            });
            filteredThree.map((x) => {
                if (m[x]) {
                    m[x] += 1;
                } else {
                    m[x] = 1;
                }
            });
            sum += Object.entries(m).filter((v) => v[1] === 3).reduce((prev, curr) => prev += this.asciiToPriority(curr[0].charCodeAt(0)), 0);
            // passed in tests but for real thing cant get this approach working so just chose to use a map...
            // logger.info(iteration);
            // const similar = this.triplePointerTrace(input[i].split(''), input[i+1].split(''), input[i+2].split(''));
            // sum += similar.reduce((prev, curr) => prev + this.asciiToPriority(curr.charCodeAt(0)), 0);
        }
        return sum;
    }
    async post(request: RucksackRequest, reply: FastifyReply) {
        logger.info('Calculating rucksack', { data: request.data });
        let sum: number;
        if (request.v2) {
            sum = this.parseGroup(request.data);
        } else {
            sum = this.parseText(request.data);
        }
        logger.info('Calculated rucksack', { sum });
        reply
            .status(200)
            .header('Content-Type', 'application/json; charset=utf-8')
            .send({ sum });
    }
}