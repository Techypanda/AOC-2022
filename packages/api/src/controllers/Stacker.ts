import { FastifyReply } from "fastify";
import { logger } from "../utils/logger"

export class SupplyStack {
    constructor(private m: Record<number, string[]> = {}) { }
    add(val: string, position: number) {
        if (this.m[position]) {
            this.m[position].unshift(val);
        } else {
            this.m[position] = [val];
        }
    }
    constructWithLength(len: number) {
        for (let i = 1; i <= len; i++) {
            if (!this.m[i]) {
                this.m[i] = [];
            }
        }
    }
    move(from: number, to: number, amount = 1) {
        const toMove: string[] = [];
        for (let i = 0; i < amount; i++) {
            toMove.unshift(this.m[from].pop() as string);
        }
        this.m[to] = [...this.m[to], ...toMove];
    }
    top(n: number) { return this.m[n][this.m[n].length - 1] }
    get length() { return Object.keys(this.m).length }
}
export interface MoveOperation {
    amount: number;
    from: number;
    to: number;
}

interface StackRequest {
    data: string[];
    v2: boolean;
}

export class Stacker {
    parse(input: string[]) {
        const stack = new SupplyStack();
        let size = -1;
        let instructionsStart = -1;
        for (let i = 0; i < input.length; i++) {
            const line = input[i];
            const row = line.split('');
            let idx = 1;
            let curr = 1;
            while (curr < line.length) {
                if (row[curr] != ' ' && isNaN(Number.parseInt(row[curr]))) {
                    stack.add(row[curr], idx);
                }
                curr += 4;
                idx += 1;
            }
            size = Math.max(size, idx - 1);
            if (line.length < 2) {
                instructionsStart = i + 1;
                break;
            }
        }
        stack.constructWithLength(size);
        const operations: MoveOperation[] = [];
        for (let i = instructionsStart; i < input.length; i++) {
            const line = input[i].split(' ');
            if (line.length === 6) {
                const operation = {
                    amount: Number.parseInt(line[1]),
                    from: Number.parseInt(line[3]),
                    to: Number.parseInt(line[5])
                }
                operations.push(operation);
            }
        }
        return { stack, operations };
    }

    async post(request: StackRequest, reply: FastifyReply) {
        logger.info('Calculating stack', { data: request.data });
        const { stack, operations } = this.parse(request.data);
        operations.forEach((op) => {
            if (request.v2) {
                stack.move(op.from, op.to, op.amount);
            } else {
                for (let i = 0; i < op.amount; i++) {
                    stack.move(op.from, op.to);
                }
            }
        });
        let str = "";
        for (let i = 1; i <= stack.length; i++) {
            str = `${str}${stack.top(i)}`;
        }
        reply
            .status(200)
            .header('Content-Type', 'application/json; charset=utf-8')
            .send({ str });
    }
}