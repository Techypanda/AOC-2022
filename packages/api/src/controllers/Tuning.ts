import { FastifyReply } from "fastify";
import { logger } from "../utils/logger";

interface TuningRequest {
    data: string;
    sequenceCount: number;
}

export class Tuning {
    detect(input: string, target = 4) {
        target -= 1;
        const chars = input.split('');
        let i = 0;
        for (i; i < chars.length - target; i++) {
            const asciiCodes: number[] = [];
            for (let k = i; k <= i + target; k++) {
                asciiCodes.push(chars[k].charCodeAt(0));
            }
            if (asciiCodes.sort((x, y) => x - y).filter((v, i, self) => self.indexOf(v) === i).length === (target+1)) {
                break;
            }
        }
        return i + target;
    }
    async post(request: TuningRequest, reply: FastifyReply) {
        logger.info('Tuning', { data: request.data, sequenceCount: request.sequenceCount });
        const packetMarker = this.detect(request.data, request.sequenceCount) + 1;
        reply
            .status(200)
            .header('Content-Type', 'application/json; charset=utf-8')
            .send({ packetMarker });
    }
}