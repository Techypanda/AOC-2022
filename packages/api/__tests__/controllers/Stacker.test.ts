import { Stacker } from "../../src/controllers/Stacker";
import { logger } from "../../src/utils/logger";

describe("Stacker works when", () => {
    const example = `    [D]    
[N] [C]    
[Z] [M] [P]
 1   2   3 

move 1 from 2 to 1
move 3 from 1 to 3
move 2 from 2 to 1
move 1 from 1 to 2`;
    it('can parse the example', () => {
        const stacker = new Stacker();
        const { stack, operations } = stacker.parse(example.split("\n"));
        expect(stack).toMatchSnapshot();
        expect(operations).toMatchSnapshot();
    });

    it('can get correct result in example 1', () => {
        const stacker = new Stacker();
        const { stack, operations } = stacker.parse(example.split("\n"));        
        operations.forEach((op) => {
            for (let i = 0; i < op.amount; i++) {
                stack.move(op.from, op.to);
            }
        });
        expect(stack).toMatchSnapshot();
        expect(stack.length).toBe(3);
        let str = "";
        for (let i = 1; i <= stack.length; i++) {
            str = `${str}${stack.top(i)}`;
        }
        expect(str).toEqual('CMZ');
    });

    it('can get correct result in example 2', () => {
        const stacker = new Stacker();
        const { stack, operations } = stacker.parse(example.split("\n"));  
        operations.forEach((op) => {
            stack.move(op.from, op.to, op.amount);
        });
        expect(stack).toMatchSnapshot();
        expect(stack.length).toBe(3);
        let str = "";
        for (let i = 1; i <= stack.length; i++) {
            str = `${str}${stack.top(i)}`;
        }
        expect(str).toEqual('MCD');
    });
})