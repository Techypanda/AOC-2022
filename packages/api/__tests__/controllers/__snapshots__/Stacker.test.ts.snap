// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stacker works when can get correct result in example 1 1`] = `
SupplyStack {
  "m": {
    "1": [
      "C",
    ],
    "2": [
      "M",
    ],
    "3": [
      "P",
      "D",
      "N",
      "Z",
    ],
  },
}
`;

exports[`Stacker works when can get correct result in example 2 1`] = `
SupplyStack {
  "m": {
    "1": [
      "M",
    ],
    "2": [
      "C",
    ],
    "3": [
      "P",
      "Z",
      "N",
      "D",
    ],
  },
}
`;

exports[`Stacker works when can parse the example 1`] = `
SupplyStack {
  "m": {
    "1": [
      "Z",
      "N",
    ],
    "2": [
      "M",
      "C",
      "D",
    ],
    "3": [
      "P",
    ],
  },
}
`;

exports[`Stacker works when can parse the example 2`] = `
[
  {
    "amount": 1,
    "from": 2,
    "to": 1,
  },
  {
    "amount": 3,
    "from": 1,
    "to": 3,
  },
  {
    "amount": 2,
    "from": 2,
    "to": 1,
  },
  {
    "amount": 1,
    "from": 1,
    "to": 2,
  },
]
`;
